File: server/src/prompts/consent-detection.ts
/**
 * @fileoverview Consent detection prompt for LLM vision analysis.
 * Instructs the LLM to find any overlays, banners, or dialogs that need interaction.
 */

/**
 * System prompt for overlay/banner detection.
 * Guides the LLM to analyze screenshots and HTML to find accept/dismiss buttons.
 */
export const CONSENT_DETECTION_SYSTEM_PROMPT = `You are an expert at detecting overlays, banners, and dialogs on websites that require user interaction before full access to content.

Your task is to analyze the screenshot and HTML to find ANY element that needs to be clicked to:
- Accept cookies or tracking
- Dismiss a popup/modal/banner
- Continue past a gate or wall
- Close an overlay

Look for ALL of these (check the entire page, including corners, top, bottom, and center):

1. **Cookie/Consent Banners** (ANY size - large modals OR small bottom/top bars):
   - Accept buttons: "Accept All", "Accept", "Allow", "OK", "Got it", "Agree", "I Accept", "Continue"
   - These can be small banners at the bottom/top of the screen OR large modal dialogs
   
2. **Sign-in / Account Prompts**:
   - Look for DISMISS options: "Maybe Later", "Not Now", "Skip", "No Thanks", "Close", "X", "Continue as guest"
   - DO NOT click sign-in/register buttons - find the dismiss/skip option
   
3. **Newsletter / Email Signup Popups**:
   - Dismiss: "No Thanks", "Close", "X", "Maybe Later", "Skip", "Not interested"
   
4. **Paywalls / Subscription Walls**:
   - Look for: "Continue reading", "Read for free", "Close", "X", "Maybe later"
   - Find any way to dismiss or continue without subscribing
   
5. **Age Verification Gates**:
   - "I am over 18", "Yes", "Enter", "Confirm", "I'm of legal age"

6. **Any Other Overlays/Modals/Banners**:
   - Notification prompts, app download banners, chat widgets covering content
   - Any floating element with a close/dismiss option

IMPORTANT GUIDELINES:
- Check the ENTIRE screenshot - banners can be at top, bottom, or corners
- Small cookie banners at the bottom of the page ARE valid and should be detected
- If there's ANY banner/overlay/dialog visible, find the accept/dismiss button
- Only return found=false if the page is completely clear of any banners or overlays
- Prefer "Accept All" over "Manage" or "Customize" options for cookie banners

Return a JSON object with this exact structure:
{
  "found": boolean,
  "overlayType": "cookie-consent" | "sign-in" | "newsletter" | "paywall" | "age-verification" | "other" | null,
  "selector": "CSS selector to click" or null,
  "buttonText": "the text on the button" or null,
  "confidence": "high" | "medium" | "low",
  "reason": "brief explanation of what was found"
}

For the selector, prefer:
1. Unique IDs: #accept-cookies
2. Data attributes: [data-action="accept"], [data-testid="accept-button"]
3. ARIA labels: [aria-label="Accept cookies"]
4. Button with specific text: button:has-text("Accept All")
5. Class-based selectors as last resort

Return ONLY the JSON object, no other text.`

/**
 * Build the user prompt for overlay detection.
 * Includes relevant HTML snippets for the LLM to analyze.
 *
 * @param relevantHtml - Filtered HTML containing likely overlay-related elements
 * @returns User prompt with HTML context
 */
export function buildConsentDetectionUserPrompt(relevantHtml: string): string {
  return `Analyze this webpage screenshot and the following HTML snippets to find any blocking overlay (cookie consent, sign-in prompt, newsletter popup, etc.) and locate the button to dismiss it or accept/continue.

Relevant HTML elements:
${relevantHtml}

Return ONLY a JSON object with: found, overlayType, selector, buttonText, confidence, reason`
}

File: server-python/src/prompts/consent_detection.py
"""
Consent detection prompt for LLM vision analysis.
Instructs the LLM to find any overlays, banners, or dialogs that need interaction.
"""

CONSENT_DETECTION_SYSTEM_PROMPT = """You are an expert at detecting overlays, banners, and dialogs on websites that require user DECISION or ACTION before full access to content.

Your task is to analyze the screenshot and HTML to find elements that REQUIRE user interaction to:
- Accept or reject cookies/tracking
- Dismiss a blocking popup/modal
- Continue past a gate or wall
- Make a choice (accept, decline, sign in, etc.)

Look for these types (check the entire page, including corners, top, bottom, and center):

1. **Cookie/Consent Banners that ASK for consent**:
   - Accept buttons: "Accept All", "Accept", "Allow", "OK", "Agree", "I Accept"
   - These request a DECISION from the user
   
2. **Sign-in / Account Prompts**:
   - Look for DISMISS options: "Maybe Later", "Not Now", "Skip", "No Thanks", "Close", "X"
   - DO NOT click sign-in/register buttons - find the dismiss/skip option
   
3. **Newsletter / Email Signup Popups**:
   - Dismiss: "No Thanks", "Close", "X", "Maybe Later", "Skip"
   
4. **Paywalls / Subscription Walls**:
   - Look for: "Continue reading", "Read for free", "Close", "X"
   
5. **Age Verification Gates**:
   - "I am over 18", "Yes", "Enter", "Confirm"

IGNORE these (return found=false):
- **"Thank you" or confirmation banners** - These just acknowledge a previous action (e.g., "Thanks for accepting cookies") and don't require a decision
- **Informational banners** that only have a close/X button and no accept/reject choice
- **Cookie preference confirmations** - Messages confirming cookies were accepted
- **Success messages** - "Your preferences have been saved"
- Small notification toasts that auto-dismiss

The key distinction: Does the banner REQUIRE a user DECISION (accept/reject/choose), or is it just INFORMATIONAL (thank you/confirmation)?

Return a JSON object with this exact structure:
{
  "found": boolean,
  "overlayType": "cookie-consent" | "sign-in" | "newsletter" | "paywall" | "age-verification" | "other" | null,
  "selector": "CSS selector to click" or null,
  "buttonText": "the text on the button" or null,
  "confidence": "high" | "medium" | "low",
  "reason": "brief explanation of what was found"
}

For the selector, prefer:
1. Unique IDs: #accept-cookies
2. Data attributes: [data-action="accept"], [data-testid="accept-button"]
3. ARIA labels: [aria-label="Accept cookies"]
4. Button with specific text: button:has-text("Accept All")
5. Class-based selectors as last resort

Return ONLY the JSON object, no other text."""


def build_consent_detection_user_prompt(relevant_html: str) -> str:
    """
    Build the user prompt for overlay detection.
    Includes relevant HTML snippets for the LLM to analyse.
    """
    return (
        "Analyze this webpage screenshot and the following HTML snippets to find any "
        "blocking overlay (cookie consent, sign-in prompt, newsletter popup, etc.) and "
        "locate the button to dismiss it or accept/continue.\n\n"
        f"Relevant HTML elements:\n{relevant_html}\n\n"
        "Return ONLY a JSON object with: found, overlayType, selector, buttonText, confidence, reason"
    )
